clang/runtime/dpct-rt/include/dpl_extras/vector.h

```diff
@@ -66,145 +213,189 @@ class device_vector {
 
   void _set_capacity_and_alloc() {
     _capacity = ::std::max(_size * 2, _min_capacity());
-    _storage = _alloc.allocate(_capacity);
+    _storage = alloc_traits::allocate(_alloc, _capacity);
+  }
+
+  void _construct(size_type n, size_type start_idx = 0) {
+    if (n > 0) {
+      device_allocator_traits<Allocator>::uninitialized_value_construct_n(
+          _alloc, _storage + start_idx, n);
+    }
+  }
+
+  void _construct(size_type n, const T &value, size_type start_idx = 0) {
+    if (n > 0) {
+      device_allocator_traits<Allocator>::uninitialized_fill_n(
+          _alloc, _storage + start_idx, n, value);
+    }
+  }
+
+  template <typename Iter>
+  void _construct(Iter first, Iter last, size_type start_idx = 0) {
+    size_type num_eles = ::std::distance(first, last);
```

Comment:
I think we can accept size as the input argument to avoid call of `std::distance` since the size is known in use positions.

Reply:
The `_size` member variable corresponds to the size of the allocation, not necessarily the number of elements to construct.  The `num_eles` we need here is the number of elements to construct.  

This is a private helper member function to help enable cases where we are constructing a sequence elements between a first and last iterator.  There are other overloads of `_construct` which accomplish what you are asking for with a known size and they are used when we have that information.  The callers of this specific overload just have the `first` and `last` iterator.  They must either use `std::distance` here or before the call. To me, it makes sense to include it here in the code.

Reply:
fixed, thanks.